### PR TITLE
Adding support for stm32c011j4m6

### PR DESCRIPTION
**Pull request by alexnagelberg**
* [ ] Adds C011J(4-6)Mxx variants

### CODE REVIEW

File: variants/STM32C0xx/C011J(4-6)M/ldscript.ld [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+_estack = ORIGIN(RAM) + LENGTH(RAM); /* end of "RAM" Ram type memory */
+
+_Min_Heap_Size = 0x200; /* required amount of heap */
+_Min_Stack_Size = 0x400; /* required amount of stack */
+
+/* Memories definition */
+MEMORY
+{
+  RAM    (xrw)    : ORIGIN = 0x20000000,   LENGTH = 6K
+  FLASH    (rx)    : ORIGIN = 0x8000000,   LENGTH = 16K
```
Comment (fpistm) [inline variants/STM32C0xx/C011J(4-6)M/ldscript.ld:48]:
```suggestion
  RAM    (xrw)    : ORIGIN = 0x20000000,   LENGTH = LD_MAX_DATA_SIZE
  FLASH    (rx)    : ORIGIN = 0x8000000+ LD_FLASH_OFFSET, LENGTH = LD_MAX_SIZE - LD_FLASH_OFFSET
```

**Review by fpistm — CHANGES_REQUESTED** [review]
Hi @alexnagelberg 
Thanks for this PR.
Could you update tour PR to reference the new generic in the README.md and add a generic clock using LL API (you can use STM32CubeMx to generate it).
Plus a small update of the ldscript to be generic.
Here an example: https://github.com/stm32duino/Arduino_Core_STM32/pull/2048/files

**Comment by alexnagelberg** [comment]
> Hi @alexnagelberg 
> 
> Thanks for this PR.
> 
> Could you update tour PR to reference the new generic in the README.md and add a generic clock using LL API (you can use STM32CubeMx to generate it).
> 
> Plus a small update of the ldscript to be generic.
> 
> Here an example: https://github.com/stm32duino/Arduino_Core_STM32/pull/2048/files

@fpistm Sure, although this ldscript was generated by cubemx, so I'm not sure how to get it to include but will play around. Just a heads up that this PR is not ready so I put it in draft, mostly as a reminder to myself to finish it.

**Comment by alexnagelberg** [comment]
Added remaining changes. Still in testing with my new chip.

File: README.md [hunk 2]
```diff
@@ -204,6 +204,7 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
 | :----: | :-------: | ---- | :-----: | :---- |
 | :green_heart: | STM32C011D6 | Generic Board | *2.6.0* |  |
 | :green_heart: | STM32C011F4<br>STM32C011F6 | Generic Board | *2.6.0* |  |
+| :yellow_heart: | STM32C011J4<br>STM32C011J6 | Generic Board | **2.7.1** |  |
```
Comment (fpistm) [inline README.md:207]:
```suggestion
| :yellow_heart: | STM32C011J4<br>STM32C011J6 | Generic Board | **2.8.0** |  |
```

File: variants/STM32C0xx/C011J(4-6)M/generic_clock.c [hunk 3]
```diff
@@ -18,10 +18,37 @@
   * @param  None
   * @retval None
   */
-WEAK void SystemClock_Config(void)
+
+void SystemClock_Config(void)
```
Comment (fpistm) [inline variants/STM32C0xx/C011J(4-6)M/generic_clock.c:22]: Keep it WEAK to allow redefinition if needed
```suggestion
WEAK void SystemClock_Config(void)
```

File: variants/STM32C0xx/C011J(4-6)M/generic_clock.c [hunk 4]
```diff
@@ -18,10 +18,37 @@
...
+
+  /** Initializes the RCC Oscillators according to the specified parameters
+  * in the RCC_OscInitTypeDef structure.
+  */
+  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSI;
+  RCC_OscInitStruct.HSIState = RCC_HSI_ON;
+  RCC_OscInitStruct.HSIDiv = RCC_HSI_DIV4;
+  RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
+  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK)
+  {
```
Comment (fpistm) [inline variants/STM32C0xx/C011J(4-6)M/generic_clock.c:35]:
```suggestion
  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
```

File: variants/STM32C0xx/C011J(4-6)M/generic_clock.c [hunk 5]
```diff
@@ -18,10 +18,37 @@
...
+    Error_Handler();
+  }
+
+  /** Initializes the CPU, AHB and APB buses clocks
+  */
+  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK|RCC_CLOCKTYPE_SYSCLK
+                              |RCC_CLOCKTYPE_PCLK1;
```
Comment (fpistm) [inline variants/STM32C0xx/C011J(4-6)M/generic_clock.c:42]:
```suggestion
  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK
                                | RCC_CLOCKTYPE_PCLK1;
```

File: variants/STM32C0xx/C011J(4-6)M/generic_clock.c [hunk 6]
```diff
@@ -18,10 +18,37 @@
...
+  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_HSI;
+  RCC_ClkInitStruct.SYSCLKDivider = RCC_SYSCLK_DIV1;
+  RCC_ClkInitStruct.AHBCLKDivider = RCC_HCLK_DIV1;
+  RCC_ClkInitStruct.APB1CLKDivider = RCC_APB1_DIV1;
+
+  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_0) != HAL_OK)
+  {
```
Comment (fpistm) [inline variants/STM32C0xx/C011J(4-6)M/generic_clock.c:49]:
```suggestion
  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_0) != HAL_OK) {
```

**Review by fpistm — CHANGES_REQUESTED** [review]
Some minor update to avoid astyle issue.

**Comment by fpistm** [comment]
Pins defined in the variant generic have no link with the built-in bootloader. Generic files could not be changed as they are automatically generated.

**Review by fpistm — APPROVED** [review]
LGTM

**Comment by fpistm** [comment]
@alexnagelberg If it is ok for you I will merge the PR.

**Comment by alexnagelberg** [comment]
> @alexnagelberg If it is ok for you I will merge the PR.

@fpistm Serial doesn't appear to be setup correctly but everything else appears working. I'm okay with it being merged and staying yellow heart.

**Comment by fpistm** [comment]
> Serial doesn't appear to be setup correctly

The only thing which could avoid serial to work is a wrong clock config anyway on C0 clock tree is not complicated so it should basically work.

**Comment by fpistm** [comment]
> > Serial doesn't appear to be setup correctly
> 
> The only thing which could avoid serial to work is a wrong clock config anyway on C0 clock tree is not complicated so it should basically work. What is the issue? wrong speed ? ...

Looking at the package, by default Serial is mapped on USART1 Using PA0 and PA1 but they are both on the same pin. So to test you can try to redefine the RX on PC14 before calling `begin()`

`Serial.setRx(PC14);`

**Comment by fpistm** [comment]
I've squash you PR and update the clock config to use LL instead of HAL. This allows to save 1K flash.

**Comment by fpistm** [comment]
Codespell issue can be safely ignored (see #2379).

**Review by fpistm — APPROVED** [review]

**Pull request merged**